File: onyx-database/src/com/onyx/diskmap/base/DiskSkipListMap.java
package com.onyx.diskmap.base;

import com.onyx.exception.AttributeMissingException;
import com.onyx.exception.AttributeTypeMismatchException;
import com.onyx.diskmap.DiskMap;
import com.onyx.diskmap.base.concurrent.EmptyReadWriteLock;
import com.onyx.diskmap.base.skiplist.AbstractIterableSkipList;
import com.onyx.diskmap.node.*;
import com.onyx.diskmap.store.Store;
import com.onyx.util.OffsetField;
import com.onyx.util.ReflectionUtil;

import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;
import java.util.function.BiConsumer;

/**
 * Created by tosborn1 on 1/7/17.
 * <p>
 * This class was added to enhance the existing index within Onyx Database.  The bitmap was very efficient but, it was a hog
 * as far as how much space it took over.  As far as in-memory data structures, this will be the go-to algorithm.  The
 * data structure is based on a SkipList.
 *
 * @param <K> Key Object Type
 * @param <V> Value Object Type
 * @since 1.2.0
 */
@SuppressWarnings("unchecked")
public class DiskSkipListMap<K, V> extends AbstractIterableSkipList<K, V> implements DiskMap<K, V> {

    private ReadWriteLock readWriteLock;

    /**
     * Constructor with store.  Initialize the collection types
     *
     * @param store  Underlying storage mechanism
     * @param header Header location of the skip list
     * @since 1.2.0
     */
    public DiskSkipListMap(Store store, Header header) {
        super(store, header);
        readWriteLock = new ReentrantReadWriteLock(true);
    }

    /**
     * Constructor with store.  Initialize the collection types
     *
     * @param store    Underlying storage mechanism
     * @param header   Header location of the skip list
     * @param detached Whether the map is headless and should ignore updating the header
     * @since 1.2.0
     */
    public DiskSkipListMap(Store store, Header header, boolean detached) {
        super(store, header, detached);

        // If it is detached.  The concurrency will be handled by this class' parent
        if (detached)
            readWriteLock = new EmptyReadWriteLock();
    }


    /**
     * Remove an item within the map
     *
     * @param key Key Identifier
     * @return The value that was removed
     */
    @SuppressWarnings("unchecked")
    @Override
    public V remove(Object key) {
        readWriteLock.writeLock().lock();

        try {
            return super.remove(key);
        } finally {
            readWriteLock.writeLock().unlock();
        }
    }

    /**
     * Put a value into a map based on its key.
     *
     * @param key   Key identifier of the value
     * @param value Underlying value
     * @return The value of the object that was just put into the map
     */
    @Override
    public V put(K key, V value) {
        readWriteLock.writeLock().lock();

        try {
            return super.put(key, value);
        } finally {
            readWriteLock.writeLock().unlock();
        }
    }

    /**
     * Get an item based on its key
     *
     * @param key Identifier
     * @return The corresponding value
     */
    @Override
    public V get(Object key) {
        readWriteLock.readLock().lock();

        try {
            return super.get(key);
        } finally {
            readWriteLock.readLock().unlock();
        }
    }

    /**
     * Iterates through the entire skip list to see if it contains the value you are looking for.
     * <p>
     * Note, this is not efficient.  It will basically do a bubble search.
     *
     * @param value Value you are looking for
     * @return Whether the value was found
     */
    @Override
    public boolean containsValue(Object value) {
        readWriteLock.readLock().lock();

        try {
            for (V next : values()) {
                if (next.equals(value))
                    return true;
            }
            return false;
        } finally {
            readWriteLock.readLock().unlock();
        }
    }

    /**
     * Put all the elements from the map into the skip list map
     *
     * @param m Map to convert from
     */
    @Override
    public void putAll(Map<? extends K, ? extends V> m) {
        m.forEach((BiConsumer<K, V>) this::put);
    }

    /**
     * Clear all the elements of the array.  If it is not detached we must handle
     * the head of teh data structure
     *
     * @since 1.2.0
     */
    @Override
    public void clear() {
        readWriteLock.writeLock().lock();

        try {

            super.clear();

            if (!this.detached) {
                setHead(createHeadNode(Byte.MIN_VALUE, 0L, 0L));
                this.header.firstNode = getHead().position;
                updateHeaderFirstNode(header, this.header.firstNode);
                header.recordCount.set(0L);
                updateHeaderRecordCount();
            }

        } finally {
            readWriteLock.writeLock().unlock();
        }
    }

    /**
     * Get the record id of a corresponding node.  Note, this points to the SkipListNode position.  Not the actual
     * record position.
     *
     * @param key Identifier
     * @return The position of the record reference if it exists.  Otherwise -1
     * @since 1.2.0
     */
    @Override
    public long getRecID(Object key) {
        readWriteLock.readLock().lock();
        try {

            SkipListNode<K> node = find((K) key);
            if (node == null)
                return -1;
            return node.position;
        } finally {
            readWriteLock.readLock().unlock();
        }
    }

    /**
     * Hydrate a record with its record ID.  If the record value exists it will be returned
     *
     * @param recordId Position to find the record reference
     * @return The value within the map
     * @since 1.2.0
     */
    @Override
    public V getWithRecID(long recordId) {
        readWriteLock.readLock().lock();

        try {
            SkipListNode<K> node = (SkipListNode<K>) findNodeAtPosition(recordId);
            return findValueAtPosition(node.recordPosition, node.recordSize);
        } finally {
            readWriteLock.readLock().unlock();
        }
    }

    /**
     * Get Map representation of key object
     *
     * @param recordId Record reference within storage structure
     * @return Map of key values
     * @since 1.2.0
     */
    public Map getMapWithRecID(long recordId) {
        readWriteLock.readLock().lock();

        try {
            SkipListNode<K> node = (SkipListNode<K>) findNodeAtPosition(recordId);
            if (node != null && node.position == recordId) {
                return getRecordValueAsDictionary(node);
            }
            return null;
        } finally {
            readWriteLock.readLock().unlock();
        }
    }

    /**
     * Get Map representation of key object.  If it is in the cache, use reflection to get it from the cache.  Otherwise,
     * just hydrate the value within the store
     *
     * @param attribute Attribute name to fetch
     * @param recordId  Record reference within storage structure
     * @return Map of key values
     * @since 1.2.0
     */
    public Object getAttributeWithRecID(final String attribute, final long recordId) throws AttributeTypeMismatchException {

        final SkipListNode node = (SkipListNode<K>) findNodeAtPosition(recordId);

        V value = findValueAtPosition(node.recordPosition, node.recordSize);
        if (value != null) {
            final Class clazz = value.getClass();
            OffsetField attributeField;
            try {
                attributeField = ReflectionUtil.getOffsetField(clazz, attribute);
            } catch (AttributeMissingException e) {
                return null;
            }
            return ReflectionUtil.getAny(value, attributeField);
        }

        return null;

    }

    @Override
    public Object getAttributeWithRecID(OffsetField attributeField, SkipListNode node) throws
            AttributeTypeMismatchException {
        if(node == null)
            return null;
        V value = findValueAtPosition(node.recordPosition, node.recordSize);

        if (value != null) {
            return ReflectionUtil.getAny(value, attributeField);
        }

        return null;
    }

    /**
     * Find all references above and perhaps equal to the key you are sending in.  The underlying data structure
     * is sorted so this should be very efficient
     *
     * @param index        The index value to compare.  This must be comparable.  It does not work with hash codes.
     * @param includeFirst Whether above and equals to
     * @return A Set of references
     * @since 1.2.0
     */
    public Set<Long> above(K index, boolean includeFirst) {
        readWriteLock.readLock().lock();

        try {

            Set results = new HashSet();
            SkipListHeadNode node = nearest(index);
            if (node != null) {

                do {

                    while (node.down != 0L)
                        node = findNodeAtPosition(node.down);

                    if (node instanceof SkipListNode) {
                        if (((SkipListNode) node).key.equals(index) && !includeFirst) {
                            if (node.next > 0L)
                                node = findNodeAtPosition(node.next);
                            else
                                break;

                            continue;
                        }

                        if (shouldMoveDown(0, 0, index, (K) ((SkipListNode) node).key))
                            results.add(((SkipListNode) node).position);
                    }
                    if (node.next > 0L)
                        node = findNodeAtPosition(node.next);
                    else
                        node = null;
                } while (node != null);
            }
            return results;
        } finally {
            readWriteLock.readLock().unlock();
        }
    }

    /**
     * Find all references below and perhaps equal to the key you are sending in.  The underlying data structure
     * is sorted so this should be very efficient
     *
     * @param index        The index value to compare.  This must be comparable.  It does not work with hash codes.
     * @param includeFirst Whether above and equals to
     * @return A Set of references
     * @since 1.2.0
     */
    public Set<Long> below(K index, boolean includeFirst) {
        readWriteLock.readLock().lock();

        try {
            Set results = new HashSet();

            SkipListHeadNode node = getHead();
            while (node.down != 0L)
                node = findNodeAtPosition(node.down);

            while (node.next != 0L) {
                node = findNodeAtPosition(node.next);
                if (node instanceof SkipListNode) {
                    if (((SkipListNode) node).key.equals(index) && !includeFirst)
                        break;
                    else if (((SkipListNode) node).key.equals(index)) {
                        results.add(node.position);
                        continue;
                    }

                    if (this.shouldMoveDown(0, 0, index, (K) ((SkipListNode) node).key))
                        break;


                    results.add(node.position);
                }
            }
            return results;
        } finally {
            readWriteLock.readLock().unlock();
        }
    }
}

File: onyx-database/src/main/kotlin/com/onyx/diskmap/impl/DiskSkipListMap.kt
package com.onyx.diskmap.impl

import com.onyx.diskmap.SortedDiskMap
import com.onyx.diskmap.impl.base.skiplist.AbstractIterableSkipList
import com.onyx.diskmap.data.Header
import com.onyx.diskmap.data.PutResult
import com.onyx.diskmap.data.SkipNode
import com.onyx.diskmap.store.Store
import com.onyx.exception.AttributeTypeMismatchException
import com.onyx.persistence.query.QueryCriteriaOperator
import com.onyx.extension.common.forceCompare
import com.onyx.extension.common.getAny
import com.onyx.lang.concurrent.ClosureReadWriteLock
import com.onyx.lang.concurrent.impl.DefaultClosureReadWriteLock
import java.lang.reflect.Field
import java.util.*

/**
 * Created by Tim Osborn on 1/7/17.
 *
 *
 * This class was added to enhance the existing index within Onyx Database.  The bitmap was very efficient but, it was a hog
 * as far as how much space it took over.  As far as in-memory data structures, this will be the go-to algorithm.  The
 * data structure is based on a SkipList.
 *
 * @param <K> Key Object Type
 * @param <V> Value Object Type
 * @since 1.2.0
 */
open class DiskSkipListMap<K, V>(fileStore:Store, header: Header, keyType:Class<*>, canStoreKeyWithinNode:Boolean) : AbstractIterableSkipList<K, V>(fileStore, header, keyType, canStoreKeyWithinNode), SortedDiskMap<K,V> {

    override val size: Int
        get() = reference.recordCount.get()

    private var mapReadWriteLock: ClosureReadWriteLock = DefaultClosureReadWriteLock()

    /**
     * Remove an item within the map
     *
     * @param key Key Identifier
     * @return The value that was removed
     */
    override fun remove(key: K): V? = mapReadWriteLock.writeLock { super.remove(key) }

    /**
     * Put a value into a map based on its key.
     *
     * @param key   Key identifier of the value
     * @param value Underlying value
     * @return The value of the object that was just put into the map
     */
    override fun put(key: K, value: V): V = mapReadWriteLock.writeLock { super.put(key, value) }

    /**
     * Put key value.  This is the same as map.put(K,V) except
     * rather than the value you just put into the map, it will
     * return the record id.  The purpose of this is so you
     * do not have to fetch the record id and search the skip list
     * again after inserting the record.
     *
     * @param key Primary Key
     * @param value Value to insert or update
     * @since 2.1.3
     * @return Value for previous record ID and if the value is been updated or inserted
     */
    override fun putAndGet(key: K, value: V, preUpdate:((Int) -> Unit)?): PutResult = mapReadWriteLock.writeLock { super.internalPutAndGet(key, value, preUpdate) }

    /**
     * Iterates through the entire skip list to see if it contains the value you are looking for.
     *
     *
     * Note, this is not efficient.  It will basically do a bubble search.
     *
     * @param value Value you are looking for
     * @return Whether the value was found
     */
    override fun containsValue(value: V): Boolean = mapReadWriteLock.readLock{
        values.forEach { next ->
            if (next == value)
                return@readLock true
        }
        return@readLock false
    }

    /**
     * Put all the elements from the map into the skip list map
     *
     * @param from Map to convert from
     */
    override fun putAll(from: Map<out K, V>) = from.forEach { this[it.key] = it.value }

    /**
     * Clear all the elements of the array.  If it is not detached we must handle
     * the head of teh data structure
     *
     * @since 1.2.0
     */
    override fun clear() = mapReadWriteLock.writeLock {
        super.clear()

        head = SkipNode.create(fileStore)
        this.reference.firstNode = head!!.position
        updateHeaderFirstNode(reference, this.reference.firstNode)
        reference.recordCount.set(0)
        updateHeaderRecordCount(0)
    }

    /**
     * Get the record id of a corresponding data.  Note, this points to the SkipListNode position.  Not the actual
     * record position.
     *
     * @param key Identifier
     * @return The position of the record reference if it exists.  Otherwise -1
     * @since 1.2.0
     */
    override fun getRecID(key: K): Int = mapReadWriteLock.readLock { find(key)?.position ?: -1  }

    /**
     * Hydrate a record with its record ID.  If the record value exists it will be returned
     *
     * @param recordId Position to find the record reference
     * @return The value within the map
     * @since 1.2.0
     */
    override fun getWithRecID(recordId: Int): V? {
        if (recordId <= 0)
            return null
        val node:SkipNode = findNodeAtPosition(recordId) ?: return null
        return node.getRecord<V>(fileStore)
    }

    /**
     * Get Map representation of key object
     *
     * @param recordId Record reference within storage structure
     * @return Map of key values
     * @since 1.2.0
     */
    override fun getMapWithRecID(recordId: Int): Map<String, Any?>? = mapReadWriteLock.readLock{
        val node:SkipNode = findNodeAtPosition(recordId) ?: return@readLock null
        return@readLock getRecordValueAsDictionary(node.record)
    }

    /**
     * Get Map representation of key object.  If it is in the cache, use reflection to get it from the cache.  Otherwise,
     * just hydrate the value within the store
     *
     * @param attribute Attribute name to fetch
     * @param reference  Record reference within storage structure
     * @return Map of key values
     * @since 1.2.0
     * @since 1.3.0 Optimized to require the reflection field so it does not have to re-instantiate one.
     */
    @Throws(AttributeTypeMismatchException::class)
    override fun <T : Any?> getAttributeWithRecID(attribute: Field, reference: Int): T = mapReadWriteLock.readLock {
        @Suppress("UNCHECKED_CAST")
        val node:SkipNode = findNodeAtPosition(reference) ?: return@readLock  null as T
        return@readLock node.getRecord<Any>(fileStore).getAny(attribute)
    }

    @Throws(AttributeTypeMismatchException::class)
    override fun <T : Any?> getAttributeWithRecID(field: Field, reference: SkipNode): T = reference.getRecord<Any>(fileStore).getAny(field)

    /**
     * Find all references above and perhaps equal to the key you are sending in.  The underlying data structure
     * is sorted so this should be very efficient
     *
     * @param index        The index value to compare.  This must be comparable.  It does not work with hash codes.
     * @param includeFirst Whether above and equals to
     * @return A Set of references
     * @since 1.2.0
     */
    override fun above(index: K, includeFirst: Boolean): Set<Int> {
        val results = HashSet<Int>()
        var node:SkipNode? = nearest(index)

        if(node != null && !node.isRecord && node.right> 0)
            node = findNodeAtPosition(node.right)

        while(node != null && node.isRecord) {
            val nodeKey:K = node.getKey(fileStore, storeKeyWithinNode, keyType)
            when {
                index.forceCompare(nodeKey) && includeFirst -> results.add(node.position)
                index.forceCompare(nodeKey, QueryCriteriaOperator.GREATER_THAN) -> results.add(node.position)
            }
            node = if (node.right > 0) findNodeAtPosition(node.right) else null
        }

        return results
    }

    /**
     * Find all references below and perhaps equal to the key you are sending in.  The underlying data structure
     * is sorted so this should be very efficient
     *
     * @param index        The index value to compare.  This must be comparable.  It does not work with hash codes.
     * @param includeFirst Whether above and equals to
     * @return A Set of references
     * @since 1.2.0
     */
    override fun below(index: K, includeFirst: Boolean): Set<Int> {
        val results = HashSet<Int>()
        var node:SkipNode? = nearest(index)

        if(node != null && !node.isRecord && node.right> 0)
            node = findNodeAtPosition(node.right)

        while(node != null && node.isRecord) {
            val nodeKey:K = node.getKey(fileStore, storeKeyWithinNode, keyType)

            when {
                index.forceCompare(nodeKey) && includeFirst -> results.add(node.position)
                index.forceCompare(nodeKey, QueryCriteriaOperator.LESS_THAN) -> results.add(node.position)
            }
            node = if (node.left > 0) findNodeAtPosition(node.left) else null
        }

        return results
    }

    /**
     * Find all references between from and to value.  The underlying data structure
     * is sorted so this should be very efficient
     *
     * @param fromValue The key to compare.  This must be comparable.  It is only sorted by comparable values
     * @param includeFrom Whether to compare above and equal or not.
     * @param toValue Key to end range to
     * @param includeTo Whether to compare equal or not.
     *
     * @since 2.1.3
     */
    override fun between(fromValue: K?, includeFrom: Boolean, toValue: K?, includeTo: Boolean): Set<Int> {
        val results = HashSet<Int>()
        var node:SkipNode? = nearest(fromValue!!)

        if(node != null && !node.isRecord && node.right> 0)
            node = findNodeAtPosition(node.right)

        node@while(node != null && node.isRecord) {
            val nodeKey:K = node.getKey(fileStore, storeKeyWithinNode, keyType)
            when {
                toValue.forceCompare(nodeKey) && includeTo -> results.add(node.position)
                !toValue.forceCompare(nodeKey, QueryCriteriaOperator.LESS_THAN) -> break@node
                fromValue.forceCompare(nodeKey) && includeFrom -> results.add(node.position)
                fromValue.forceCompare(nodeKey, QueryCriteriaOperator.GREATER_THAN) -> results.add(node.position)
            }
            node = if (node.right > 0) findNodeAtPosition(node.right) else null
        }

        return results
    }

}
